mkosi/kmod.py

```diff
@@ -55,6 +55,39 @@ def filter_kernel_modules(
     return sorted(modules)
 
 
+def filter_firmware(
+    root: Path,
+    firmware: set[Path],
+    *,
+    include: Iterable[str],
+    exclude: Iterable[str],
+) -> list[Path]:
+    firmwared = Path("usr/lib/firmware")
+    with chdir(root):
+        all_firmware = set(firmwared.rglob("*"))
+
+    keep = set()
+    if include:
+        regex = re.compile("|".join(include))
+        for f in all_firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if regex.search(rel):
+                keep.add(rel)
+                firmware.add(f)
+
+    if exclude:
+        remove = set()
+        regex = re.compile("|".join(exclude))
+        for f in firmware:
+            rel = os.fspath(Path(*f.parts[3:]))
+            if rel not in keep and regex.search(rel):
+                remove.add(f)
+
+        firmware -= remove
+
+    return sorted(firmware)
```

Comment:
Sorting here when we're immediately going to turn it into an (unordered) set afterwards is just wasteful. Let's immediately return a set from this function instead.